### PR TITLE
Jwoo/explicit image size uwp

### PR DESCRIPTION
**Pull request by nesalang**
UWP changes
~~I assumed that UWP content mode of displaying image is simmilar to Scale-To-AspectRatio of iOS, so I set only the dimension that has unsigned non-zero values when one them is zero.~~
Allows image size to be explicitly set given size (width and height) even if doing so will distort the image.
- If only one dimension is given, the given dimension is explicitly set, and the missing dimension will be adopted to keep the aspect ratio of the image.
- takes unsigned integer followed by unit; currently only support pix ("px").

### CODE REVIEW

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 1]
```diff
@@ -1471,6 +1471,11 @@ AdaptiveNamespaceStart
         ComPtr<IAdaptiveCardResourceResolvers> resourceResolvers;
         THROW_IF_FAILED(renderContext->get_ResourceResolvers(&resourceResolvers));
 
+        UINT32 explicitWidth = 0, explicitHeight = 0;
+        THROW_IF_FAILED(adaptiveImage->get_Width(&explicitWidth));
+        THROW_IF_FAILED(adaptiveImage->get_Height(&explicitHeight));
+        bool isAspectRaitioNeeded = (explicitWidth  && explicitHeight);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >Raitio [](start = 21, length = 6)

Spelling (Ratio)

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 2]
```diff
@@ -1480,9 +1485,15 @@ AdaptiveNamespaceStart
             ComPtr<IShape> ellipseAsShape;
             THROW_IF_FAILED(ellipse.As(&ellipseAsShape));
             // Set Auto, None, and Stretch to Stretch_UniformToFill. An ellipse set to Stretch_Uniform ends up with size 0.
-            if (size == ABI::AdaptiveNamespace::ImageSize::None ||
+            if (isAspectRaitioNeeded)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: Could you add a test file to samples\Tests that demonstrates this and run the test app?

File: samples/v1.0/Elements/Image.Explicit.Size.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
+{
```
Comment (RebeccaAnne) [inline samples/v1.0/Elements/Image.Explicit.Size.json]: Can we add a person type example as well to exercise the ellipse code path?

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 4]
```diff
@@ -1525,6 +1536,11 @@ AdaptiveNamespaceStart
             SetImageOnUIElement(imageUri.Get(), xamlImage.Get(), resourceResolvers.Get());
             THROW_IF_FAILED(xamlImage.As(&frameworkElement));
 
+            if(isAspectRaitioNeeded)
+            {
+                xamlImage->put_Stretch(Stretch::Stretch_Fill);
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >                [](start = 62, length = 15)

[nit] remove trailing spaces

File: samples/v1.0/Elements/Image.Explicit.Size.json (hunk 3 above)
Comment (nesalang) [reply samples/v1.0/Elements/Image.Explicit.Size.json]: sure

---
In reply to: [184109786](https://github.com/Microsoft/AdaptiveCards/pull/1373#discussion_r184109786) [](ancestors = 184109786)

File: samples/v1.0/Elements/Image.Explicit.Size.json [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+			"type": "TextBlock",
+			"text": "width:50px height:50px",
+			"weight": "bolder",
+			"size": "large"
+		},
+		{
+			"type": "Image",
+			"url": "http://adaptivecards.io/content/cats/1.png",
+			"size": "small",
+                        "width": "50px",
```
[the remark this answers is not in the document]
Comment (nesalang) [reply samples/v1.0/Elements/Image.Explicit.Size.json]: thanks

---
In reply to: [184108904](https://github.com/Microsoft/AdaptiveCards/pull/1373#discussion_r184108904) [](ancestors = 184108904)

File: source/uwp/Renderer/lib/XamlBuilder.cpp [hunk 6]
```diff
@@ -1487,20 +1491,31 @@ AdaptiveNamespaceStart
...
             THROW_IF_FAILED(ellipse.As(&ellipseAsShape));
+
             // Set Auto, None, and Stretch to Stretch_UniformToFill. An ellipse set to Stretch_Uniform ends up with size 0.
             if (size == ABI::AdaptiveNamespace::ImageSize::None ||
                 size == ABI::AdaptiveNamespace::ImageSize::Stretch ||
-                size == ABI::AdaptiveNamespace::ImageSize::Auto)
+                size == ABI::AdaptiveNamespace::ImageSize::Auto ||
+                explicitWidth ||
+                explicitHeight ||
+                isAspectRatioNeeded)
```
Comment (RebeccaAnne) [inline source/uwp/Renderer/lib/XamlBuilder.cpp]: >isAspectRatioNeeded [](start = 16, length = 19)

Don't need this, if this is true, explicitWidth and explicitHeight will already be true.

**Pull request merged**